token/program-2022/src/extension/mod.rs

```diff
@@ -1133,6 +1136,7 @@ impl ExtensionType {
                 }
                 ExtensionType::NonTransferable => {
                     account_extension_types.push(ExtensionType::NonTransferableAccount);
+                    account_extension_types.push(ExtensionType::ImmutableOwner);
```

Comment:
The changes here are a bit invasive, since it's going further than checking the extensions at init, and rather adding the immutable ownership extension by force for all non-transferable accounts. This should be fine, but let me think about it some more.